File: Household Budget Calculator/[VB]-Household Budget Calculator/VB/Household budget calculator/expanding property classes/RegularBills.vb
Imports System.ComponentModel

<TypeConverter(GetType(ExpandableObjectConverter))> <System.Serializable()> _
Public Class RegularBills

    Public Event propertiesChanged(ByVal e As propertiesChangedEventArgs)

    Private _councilTax As Decimal
    <DisplayName("Council Tax"), _
    RefreshProperties(RefreshProperties.All)> _
    Public Property councilTax() As Decimal
        Get
            Return _councilTax
        End Get
        Set(ByVal value As Decimal)
            _councilTax = value
            totalRegularBills = councilTax + amenities + communications
        End Set
    End Property

    Private _amenities As Decimal
    <DisplayName("Gas/Electricity/Water"), _
    RefreshProperties(RefreshProperties.All)> _
    Public Property amenities() As Decimal
        Get
            Return _amenities
        End Get
        Set(ByVal value As Decimal)
            _amenities = value
            totalRegularBills = councilTax + amenities + communications
        End Set
    End Property

    Private _communications As Decimal
    <DisplayName("Phone/Television"), _
    RefreshProperties(RefreshProperties.All)> _
    Public Property communications() As Decimal
        Get
            Return _communications
        End Get
        Set(ByVal value As Decimal)
            _communications = value
            totalRegularBills = councilTax + amenities + communications
        End Set
    End Property

    Private _totalRegularBills As Decimal
    <ReadOnlyAttribute(True), _
    Browsable(False), _
    DisplayName("Total Regular Bills Outgoings")> _
    Public Property totalRegularBills() As Decimal
        Get
            Return _totalRegularBills
        End Get
        Set(ByVal value As Decimal)
            _totalRegularBills = value
            RaiseEvent propertiesChanged(New propertiesChangedEventArgs With {.propName = "Regular Bills", .newValue = value})
        End Set
    End Property

    Public Overrides Function ToString() As String
        Return totalRegularBills.ToString("c2")
    End Function

End Class

File: Household Budget Calculator/[C#]-Household Budget Calculator/C#/Household budget calculator CS/expanding property classes/RegularBills.cs
//using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.ComponentModel;

namespace Household_budget_calculator_CS.expanding_property_classes
{
    [TypeConverter(typeof(ExpandableObjectConverter))]
    [System.Serializable()]
    public class RegularBills
    {
        public event propertiesChangedEventHandler propertiesChanged;
        public delegate void propertiesChangedEventHandler(propertiesChangedEventArgs e);

        private decimal _councilTax;
        [DisplayName("Council Tax"), RefreshProperties(RefreshProperties.All)]
        public decimal councilTax
        {
            get { return _councilTax; }
            set
            {
                _councilTax = value;
                totalRegularBills = councilTax + amenities + communications;
            }
        }

        private decimal _amenities;
        [DisplayName("Gas/Electricity/Water"), RefreshProperties(RefreshProperties.All)]
        public decimal amenities
        {
            get { return _amenities; }
            set
            {
                _amenities = value;
                totalRegularBills = councilTax + amenities + communications;
            }
        }

        private decimal _communications;
        [DisplayName("Phone/Television"), RefreshProperties(RefreshProperties.All)]
        public decimal communications
        {
            get { return _communications; }
            set
            {
                _communications = value;
                totalRegularBills = councilTax + amenities + communications;
            }
        }

        private decimal _totalRegularBills;
        [ReadOnlyAttribute(true), Browsable(false), DisplayName("Total Regular Bills Outgoings")]
        public decimal totalRegularBills
        {
            get { return _totalRegularBills; }
            set
            {
                _totalRegularBills = value;
                if (propertiesChanged != null)
                {
                    propertiesChanged(new propertiesChangedEventArgs
                    {
                        propName = "Regular Bills",
                        newValue = value
                    });
                }
            }
        }

        public override string ToString()
        {
            return totalRegularBills.ToString("c2");
        }

    }
}
